tokio/src/sync/batch_semaphore.rs

```diff
@@ -0,0 +1,509 @@
+            // Was the waiter assigned enough permits to wake it?
+            if !waiter.assign_permits(&mut rem) {
+                break;
+            }
+            last = Some(NonNull::from(waiter));
+        }
+
+        // If we assigned permits to all the waiters in the queue, and there are
+        // still permits left over, assign them back to the semaphore.
+        if rem > 0 {
+            self.permits.fetch_add(rem << Self::PERMIT_SHIFT, Release);
+        }
+
+        // Split off the queue at the last waiter that was satisfied, creating a
+        // new list. Once we release the lock, we'll drain this list and notify
+        // the waiters in it.
+        if let Some(waiter) = last {
+            // Safety: it's only safe to call `split_back` with a pointer to a
+            // node in the same list as the one we call `split_back` on. Since
+            // we got the waiter pointer from the list's iterator, this is fine.
+            unsafe { waiters.queue.split_back(waiter) }
+        } else {
+            LinkedList::new()
+        }
+    }
+
+    fn poll_acquire(
+        &self,
+        cx: &mut Context<'_>,
+        num_permits: u16,
+        node: Pin<&mut Waiter>,
+        queued: bool,
+    ) -> Poll<Result<(), AcquireError>> {
+        let mut acquired = 0;
+
+        let needed = if queued {
+            node.state.load(Acquire) << Self::PERMIT_SHIFT
+        } else {
+            (num_permits as usize) << Self::PERMIT_SHIFT
+        };
+
+        let mut lock = None;
+        // First, try to take the requested number of permits from the
+        // semaphore.
+        let mut curr = self.permits.load(Acquire);
+        let mut waiters = loop {
+            // Has the semaphore closed?
+            if curr & Self::CLOSED > 0 {
+                return Ready(Err(AcquireError::closed()));
+            }
+
+            let mut remaining = 0;
+            let (next, acq) = if curr + acquired >= needed {
+                let next = curr - (needed - acquired);
+                (next, needed >> Self::PERMIT_SHIFT)
+            } else {
+                remaining = (needed - acquired) - curr;
+                (0, curr >> Self::PERMIT_SHIFT)
+            };
+
+            if remaining > 0 && lock.is_none() {
+                // No permits were immediately available, so this permit will
+                // (probably) need to wait. We'll need to acquire a lock on the
+                // wait queue before continuing. We need to do this _before_ the
+                // CAS that sets the new value of the semaphore's `permits`
+                // counter. Otherwise, if we subtract the permits and then
+                // acquire the lock, we might miss additional permits being
+                // added while waiting for the lock.
+                lock = Some(self.waiters.lock().unwrap());
+            }
+
+            match self.permits.compare_exchange(curr, next, AcqRel, Acquire) {
```

Comment:
> Would it still be fair because as long as other waiters exist the number of available permits is always 0 (freed permits directly get added to the wait list)?

That's correct; the semaphore's permit counter will only be incremented if the wait list is empty.